### DISCUSSION TITLE
[Camera Support]: Tapo cameras audio stuttering in recordings only

### DISCUSSION THREAD

**Discussion by Ferooo00oo** [blakeblackshear/frigate#14233] [Camera Support] [answered]
### Describe the problem you are having

Tapo cameras audio stuttering in recordings only

### Version

0.14.1-

### Frigate config file

```yaml
mqtt:
  enabled: false

go2rtc:
  streams:
    Office:
      - rtsp://user:pass@192.168.1.xxx:554/stream1
    Hall:
      - rtsp://user:pass@192.168.1.yyy:554/stream1


cameras:

  Office:
    ffmpeg:
      output_args:
        record: preset-record-generic-audio-aac
        
      inputs:
        - path: rtsp://127.0.0.1:8554/Office
          input_args: preset-rtsp-restream
          roles:
            - record
        - path: rtsp://user:pass@192.168.1.xxx:554/stream2
          roles:
            - detect
    onvif:
      host: 192.168.1.xxx
      port: 2020
      user: user
      password: pass

    detect:
      width: 1920
      height: 1080
      fps: 5

    record: # <----- Enable recording
      enabled: true
      retain:
        days: 3
        mode: all





  Hall: # <--- this will be changed to your actual camera later

    ffmpeg:
      output_args:
        record: preset-record-generic-audio-aac

      inputs:
        - path: rtsp://127.0.0.1:8554/Hall
          input_args: preset-rtsp-restream
          roles:
            - record
        - path: rtsp://user:pass@192.168.1.yyy:554/stream2
          roles:
            - detect

    onvif:
      host: 192.168.1.yyy
      port: 2020
      user: user
      password: pass

    detect:
      width: 1920
      height: 1080
      fps: 5

    record: # <----- Enable recording
      enabled: true
      retain:
        days: 3
        mode: all



detectors:
  ov:
    type: openvino
    device: CPU
    model:
      path: /openvino-model/FP16/ssdlite_mobilenet_v2.xml
model:
  width: 300
  height: 300
  input_tensor: nhwc
  input_pixel_format: bgr
  labelmap_path: /openvino-model/coco_91cl_bkgr.txt
version: 0.14
camera_groups:
  VIP:
    order: 1
    icon: LuServer
    cameras:
      - Hall
      - Office
```

### Relevant Frigate log output

```shell
2024-10-09 12:32:02.942234  [INFO] Preparing Frigate...
2024-10-09 12:32:02.942274  [INFO] Starting Frigate...
2024-10-09 12:32:03.793512  [2024-10-09 12:32:03] frigate.app                    INFO    : Starting Frigate (0.14.1-)
2024-10-09 12:32:03.793558  [2024-10-09 12:32:03] frigate.util.config            INFO    : Checking if frigate config needs migration...
2024-10-09 12:32:03.809026  [2024-10-09 12:32:03] frigate.util.config            INFO    : frigate config does not need migration...
2024-10-09 12:32:05.409910  [2024-10-09 12:32:05] frigate.util.services          WARNING : Did not detect hwaccel, using a GPU for accelerated video decoding is highly recommended
2024-10-09 12:32:05.415185  [2024-10-09 12:32:05] frigate.config                 WARNING : Customizing more than a detector model path is unsupported.
2024-10-09 12:32:05.422159  [2024-10-09 12:32:05] peewee_migrate.logs            INFO    : Starting migrations
2024-10-09 12:32:05.430520  [2024-10-09 12:32:05] peewee_migrate.logs            INFO    : There is nothing to migrate
2024-10-09 12:32:05.691349  [2024-10-09 12:32:05] frigate.app                    INFO    : Recording process started: 190943
2024-10-09 12:32:05.694720  [2024-10-09 12:32:05] frigate.app                    INFO    : Recording process started: 190945
2024-10-09 12:32:05.697019  [2024-10-09 12:32:05] frigate.app                    INFO    : go2rtc process pid: 144
2024-10-09 12:32:05.713167  [2024-10-09 12:32:05] frigate.app                    INFO    : Output process started: 190972
2024-10-09 12:32:05.719334  [2024-10-09 12:32:05] detector.ov                    INFO    : Starting detection process: 190971
2024-10-09 12:32:05.740558  [2024-10-09 12:32:05] frigate.app                    INFO    : Camera processor started for Office: 190987
2024-10-09 12:32:05.743015  [2024-10-09 12:32:05] frigate.app                    INFO    : Camera processor started for Hall: 190989
2024-10-09 12:32:05.746974  [2024-10-09 12:32:05] frigate.app                    INFO    : Capture process started for Office: 190995
2024-10-09 12:32:05.753284  [2024-10-09 12:32:05] frigate.app                    INFO    : Capture process started for Hall: 190998
```

### FFprobe output from your camera

```shell
N/A
```

### Operating system

Proxmox

### Install method

Proxmox via TTeck Script

### Object Detector

CPU (no coral)

### Network connection

Wired

### Camera make and model

Tapo C100 & Tapo C200

### Screenshots of the Frigate UI's System metrics pages

![image](https://github.com/user-attachments/assets/cdeea016-0293-46dc-82d7-4a33e592a78d)
![image](https://github.com/user-attachments/assets/c8ba3288-874d-4fff-a5bc-e402c93c8a5a)

**Comment by NickM-27** [comment]
Might be worth trying to set `"ffmpeg:rtsp..."` for the go2rtc streams

**Reply by Ferooo00oo** [reply]
Thanks for the reply **-I appreciate it-** but unfortunately that didn't work.  **:(**

**Reply by NickM-27** [reply]
then it seems tapo is just sending bad quality audio, you are already transcoding it so not sure there is much that can be done

**Comment by cobraycloud** [comment]
Just wanted to say ive got the same problem with more that one tapo camera type.

**Reply by cvroque** [reply]
Hello, check if my config below also fixes it for you!

**Comment by cvroque** [comment, marked as answer]
Don't ask me to explain why, but after some trial and error I managed to get audio recording in decent quality. These are the relevant settings:

```
ffmpeg:
  output_args:
    record: preset-record-generic-audio-aac                 # This will transcode the original audio to aac

go2rtc:
  streams:
    main:
      - tapo://admin:hash_password@192.168.1.132            # Source 1: Use this for two way audio 
      - rtsp://user:password@192.168.1.132/stream1          # Source 2: Use this if first one fails, remove the first
      - "ffmpeg:main#audio=aac"                             
      - "ffmpeg:main#audio=opus"                            
    main_sd:
      - tapo://admin:hash_password@192.168.1.132?subtype=1  # Source 1: Use this for two way audio
      - rtsp://user:password@192.168.1.132/stream2          # Source 2: Use this if first one fails, remove the first

cameras:
  main:
    ffmpeg:
      inputs:
        - path: rtsp://127.0.0.1:8554/main?video&audio
          input_args: preset-rtsp-restream 
          roles:
            - record                                        # Use the main stream to record in high quality
        - path: rtsp://127.0.0.1:8554/main_sd
          input_args: preset-rtsp-restream
          roles:
            - detect                                        # Use the lower quality stream for detection
        - path: rtsp://127.0.0.1:8554/main_sd?audio 
          roles:
            - audio                                         # Just audio, if you're running sound detection  
```

I use a different source in order to get two way audio working, but should be the same settings for the RTSP source. You should use just one of the sources at the same time: [Tapo](https://github.com/AlexxIT/go2rtc#source-tapo) or [RTSP](https://github.com/AlexxIT/go2rtc?tab=readme-ov-file#source-rtsp).

For some reason asking a mp4 or aac stream from go2rtc doesn't yield good results, but the transcoding is necessary to use two way audio within Frigate.

**Reply by Ferooo00oo** [reply]
THANK YOU SO MUCH !
Your copy of the config unfortunatly didn't work for me. But when I used "rtsp://user:password@ipadress:554/stream1" instead of "tapo://" and got rid of the "audio role" part everything worked!
I appreciate it !❤️

**Reply by cvroque** [reply]
Hey, sorry for the confusion I just noticed that I removed a bit too much from the example config, that audio role was using the lower quality stream. About the "tapo://" source, the manufacturer blocked this direct access on newer firmware versions so "rtsp://" is the appropriate backup.

Since it worked for you, I edited my original comment to make it easier if anyone else has the same issue.

**Reply by Ferooo00oo** [reply]
One more question. How do you use the 2-way audio? Is it through the frigate interface or do you have another way?

**Reply by NickM-27** [reply]
yes, the frigate interface supports two way talk assuming you are accessing it via https

**Reply by cobraycloud** [reply]
> Don't ask me to explain why, but after some trial and error I managed to get audio recording in decent quality. These are the relevant settings:
> 
> ```
> ffmpeg:
>   output_args:
>     record: preset-record-generic-audio-aac                 # This will transcode the original audio to aac
> 
> go2rtc:
>   streams:
>     main:
>       - tapo://admin:hash_password@192.168.1.132            # Source 1: Use this for two way audio 
>       - rtsp://user:password@192.168.1.132/stream1          # Source 2: Use this if first one fails, remove the first
>       - "ffmpeg:main#audio=aac"                             
>       - "ffmpeg:main#audio=opus"                            
>     main_sd:
>       - tapo://admin:hash_password@192.168.1.132?subtype=1  # Source 1: Use this for two way audio
>       - rtsp://user:password@192.168.1.132/stream2          # Source 2: Use this if first one fails, remove the first
> 
> cameras:
>   main:
>     ffmpeg:
>       inputs:
>         - path: rtsp://127.0.0.1:8554/main?video&audio
>           input_args: preset-rtsp-restream 
>           roles:
>             - record                                        # Use the main stream to record in high quality
>         - path: rtsp://127.0.0.1:8554/main_sd
>           input_args: preset-rtsp-restream
>           roles:
>             - detect                                        # Use the lower quality stream for detection
>         - path: rtsp://127.0.0.1:8554/main_sd?audio 
>           roles:
>             - audio                                         # Just audio, if you're running sound detection  
> ```
> 
> I use a different source in order to get two way audio working, but should be the same settings for the RTSP source. You should use just one of the sources at the same time: [Tapo](https://github.com/AlexxIT/go2rtc#source-tapo) or [RTSP](https://github.com/AlexxIT/go2rtc?tab=readme-ov-file#source-rtsp).
> 
> For some reason asking a mp4 or aac stream from go2rtc doesn't yield good results, but the transcoding is necessary to use two way audio within Frigate.

The thing that did it for me was either the ?video&audio or the "ffmpeg:main#audio=opus"  That worked perfect!! Thanks!

**Reply by RobinBeismann** [reply]
@cvroque do you use 2 way audio nowadays? The tapo protocol is supported again since a while.
If yes, mind sharing your config?
